File: LeetCode/17 M5 - Leetcode 628 - Maximum Product of 3 Numbers/O(n)_without_sort/heap/628.py
# src: https://leetcode.com/problems/maximum-product-of-three-numbers/solution/304829

import heapq

class Solution:
    def maximumProduct(self, array):
        largest = heapq.nlargest(3, array)
        smallest = heapq.nsmallest(2,array)

        return max(largest[0] * largest[1] * largest[2], largest[0] * smallest[0] * smallest[1] )


# nlargest is O(n log k) for selecting k elements: the most efficient way
# as k is either 2 or 3, this is O(n)
# https://github.com/python/cpython/blob/2.7/Lib/heapq.py#L203

File: LeetCode/17 M5 - Leetcode 628 - Maximum Product of 3 Numbers/O(nlon)/628.cpp


#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;

class Solution {
public:

	int maximumProduct(vector<int>& nums) {
		sort(nums.begin(), nums.end());
		int n = nums.size();
		int a = nums[n - 1] * nums[n - 2] * nums[n - 3];
		int b = nums[0] * nums[1] * nums[n - 1];
		return max(a, b);
	}


};

int main() {
	Solution s;
	vector<int> nums { -20, -10, -7, -6, -5, 2, 3, 4, 5};
	cout << s.maximumProduct(nums);
}

